### ISSUE TITLE
DandelionFilter, Spring WebMVC, and Jetty

### ISSUE THREAD

**Issue by fbristow** [dandelion/dandelion#91]
Hi Thibault,
I've encountered a bug that may be difficult to reproduce, but I'm mostly certain that I've nailed down where the bug is happening.

Basically, the problem that we're encountering is that we randomly (more often than not) get connection reset errors on assets that might be handled by `DandelionFilter`, but only when we run Jetty, and only when we make a request from a non-`localhost` browser.

The problem appears to be that `DandelionFilter` allows the filter chain to proceed, and Spring handles the response for a file that's going to be served. When Spring handles the request, in `ResourceHttpRequestHandler#writeContent` (see: https://github.com/spring-projects/spring-framework/blob/master/spring-webmvc/src/main/java/org/springframework/web/servlet/resource/ResourceHttpRequestHandler.java#L442) it's writing the content out via the `OutputStream` for the `ServletResponse`. Later, `DandelionFilter` tries to continue writing headers out to the response, even though the response has been marked as completed by Spring.

I *think* that `DandelionFilter` is oblivious to other filters writing out to the response by the `OutputStream` (`ByteArrayResponseWrapper` is assuming that the response is going to be written by a `PrintWriter`).

We're using Dandelion 1.1.1.

Thanks!

**Comment by tduchateau** [comment]
Hi Franklin. Thanks for reporting, and for your investigation.
Please precise the version of Jetty you're using. I do need to reproduce this kind of issue.
Thanks

**Comment by fbristow** [comment]
Hi Thibault,
We're using Jetty 9.2.1.v20140609.

Thanks!

**Comment by tduchateau** [comment]
Indeed, quite hard to reproduce. :-/
What version of Spring? JSP/Thymeleaf?
Is there any other filter being used?

**Comment by fbristow** [comment]
Since this is such a difficult to reproduce bug (as I expected it would be) I'd like to try setting up a minimal test case that I can attach to this issue. It might take a couple days, but I'll try to get one up soon.

**Comment by tduchateau** [comment]
:+1:

**Comment by fbristow** [comment]
Hi Thibault,
I've reliably recreated the problem in a pretty minimal package, which you can find here: https://github.com/fbristow/dandelion-springmvc

The problem **does not** happen when you browse to this on `localhost`. I have only been able to reproduce the error when browsing to the test application from a different machine. The OS and browser of the other machines do not seem to make a difference, but Chrome reports `net::ERR_CONNECTION_RESET` when the bug happens. Firefox's network panel reports 200 OK and actually shows the content in the response, but it's clear that something has messed up because the resource requested isn't actually available for the page to use (for example, angular isn't available).

What I've figured out so far is that when `DandelionFilter` passes off to the rest of the filter chain (here: https://github.com/dandelion/dandelion/blob/master/dandelion-core/src/main/java/com/github/dandelion/core/web/DandelionFilter.java#L128) the request is eventually passed through to Spring's `ResourceHttpRequestHandler`.

`ResourceHttpRequestHandler` straight up copies the resource to the `OutputStream` of the response and Spring's utility class method `StreamUtils#copy` appears to close the `OutputStream` when it's finished (see here: https://github.com/spring-projects/spring-framework/blob/master/spring-webmvc/src/main/java/org/springframework/web/servlet/resource/ResourceHttpRequestHandler.java#L461) [Note: `in.close()` is closing the `InputStream` on the file that it's writing. I found that the `HttpServletResponse` output stream is being closed when `StreamUtils#copy` calls `OutputStream#flush()` when I observe the behaviour in Eclipse's debugger.]

So now, when we finally get back to `DandelionFilter` and proceed, an `EofException` is thrown by Jetty when `DandelionFilter` attempts to write out to the `OutputStream` (here: https://github.com/dandelion/dandelion/blob/master/dandelion-core/src/main/java/com/github/dandelion/core/web/DandelionFilter.java#L154)

As far as I can tell, this *only* affects Jetty. We've been running this app in production on Tomcat for a couple of months with Dandelion and we haven't encountered the same problem.

**Comment by tduchateau** [comment]
Hi Franklin,

Most of all, a huge THANKS for what you did. After hours of struggling with Jetty, I've found out that using the current implementation of response wrapper, something within the filter chain was causing the response to be commited before trying to write to the response output stream, thus leading to a complete mess. I'm still not sure to understand what exactly caused it but after some refactoring in the response wrapper, I've got much better results.
If you're OK with that, I'de love you to test the latest dandelion-core snapshot with your sample, as soon as it is available in Central.

**Comment by tduchateau** [comment]
Hi Franklin. A 1.2.0-SNAPSHOT version is available in Central. Could you please give it a try?

**Comment by fbristow** [comment]
Hi Thibault,
I tested it out, and the problem that I was experiencing is resolved. Unfortunately, this solution breaks other behaviour for me: as part of a dynamic file download feature, I will write large volumes (many gigabytes) of data out to the `OutputStream` from the response. This solution causes the file downloads to consume lots of memory on the server before they're written out, and the client browser hangs waiting for the download to start. This is something we've encountered before (see here: http://dandelion-forum.48353.x6.nabble.com/Dandelion-not-flushing-response-buffer-until-complete-td1911.html)

Would it be possible for `DandelionFilter` to simply check and see if the `OutputStream` is still open for writing before attempting to write to it?

**Comment by tduchateau** [comment]
Hi,

> Would it be possible for DandelionFilter to simply check and see if the OutputStream is still open for writing before attempting to write to it?
I think it could. But it's not so easy. And I am not very enthusiastic about wrapping things again and again. 

This is an Interesting use case however. Could you please update your sample app (https://github.com/fbristow/dandelion-springmvc) with a test case? That would help a lot to debug what's wrong.

But... Let's step back a little: you are writing gigabytes of data to the response?!? :-)

**Comment by fbristow** [comment]
Hi Thibault,
I've updated my test project to include a link that "downloads" a large file (just writes out 1GB worth of numbers).

> But... Let's step back a little: you are writing gigabytes of data to the response?!? :-)

Hehe :)

Not *quite* as crazy as it sounds! We're building an application that serves as a storage and analytical platform for genetic sequencing data. Sequencing data itself is pretty big, most of the files we're managing are in the hundreds of MB. One of the features we've developed allows end-users to select many files and download them to their local computer. Since we don't know what files any user is going to want at startup time, we dynamically build the zip file and send it back to the browser using the `OutputStream` from the `HttpServletResponse`.

**Comment by tduchateau** [comment]
> I was thinking about calling isCommitted() on the HttpServletResponse...

The actual problem comes from the response wrapper. Since the latest snapshot, the response output stream is now entirely buffered, which causes the extra memory consumption when downloading large files. This is clearly not the right solution is such cases.

The previous implementation was using a wrapper (based on a PrintWriter) over the output stream which was not buffered. I should have try to fix it first (for Jetty) before switching to a buffer-based one. And finally, you've made me realize that a response wrapper based on a buffer was not the right direction. The ideal solution would be to be able to pick only the relevant requests (the ones of interest to Dandelion) and leave all others untouched but unfortunately, there is not other choice (at least to me :-)) but to wrap the response and (try to) make it work with all app servers.
Well, let's start again from the previous ByteArrayResponseWrapper and see how it can be fixed. And thanks again for the sample app!

> We're building an application that serves as a storage and analytical platform for genetic sequencing data. Sequencing data itself is pretty big, most of the files we're managing are in the hundreds of MB

Pretty interesting stuff! :-)
I suppose you also need some gzipping to save bandwidth, don't you? If so, what are you using? Another filter? Something else?

**Comment by fbristow** [comment]
> The actual problem comes from the response wrapper.
Yep! :) I meant `isCommitted()` as a solution without wrapping the `OutputStream`.

> I suppose you also need some gzipping to save bandwidth, don't you? If so, what are you using? Another filter? Something else?

Since we're building a collection of files to send back to the client, we're wrapping the `OutputStream` from the response in a [`ZipOutputStream`](http://docs.oracle.com/javase/8/docs/api/index.html?java/util/zip/ZipOutputStream.html) and writing to that. We can set the compression level on the stream manually. We are using a `gzip` filter for serving web application files (JavaScript, CSS, etc.)

**Comment by gregw** [comment]
Hi guys - I'm from the jetty team and just a note to say please do ask us for help with any issues like this.  We are very willing to help with resolving differences between containers.

I'm guessing here, as I've not looked at the code, but I think the issue may be due to an optimisation in jetty that is causing the response to be committed before it is in Tomcat.        In Jetty, writes are mostly aggregated into a buffer, until such time as that buffer overflows or somebody calls flush.

However, we have an optimization to avoid copying large writes.   If we have a 32KB write aggregation buffer and we receive a write of 32K-1 bytes, we don't want to copy that buffer only to flush it as soon as a single extra byte is written.    So we have a commitSize, which by default is 1/4 of our buffer size.  Any write that is larger than our commit size is just wrapped and directly written (committed) without an extra copy.     

This optimization can result in Jetty committed response before other containers do.  If your filter then tries to write headers after commit, you will get a problem.

The work around i that you can call `response.setBufferSize(int size)` to explicitly set a buffer size, and because you have been explicit, we also set the commit size to the same size.      If you don't know what size to set you can even just do `response.setBufferSize(response.getBufferSize())` to just set the commit size.

In general, rather than your app/filter buffering data, it is probably better to set a response buffer size and let the container do the buffering for you as we can then more efficiently flush it without extra copies.

Also, trying to write headers after any data has been written is always going to be size dependent and thus fragile.  Best to write your headers first.
